rpc/src/responses/json_array.rs

```diff
@@ -1,47 +1,280 @@
 use super::ParseError;
 use serde::{de, Deserialize, Serialize};
-use serde_json::{self, json, Value};
+use serde_json::{self, Value};
+use std::fmt;
+use std::iter;
 
-#[derive(Serialize, Deserialize)]
+#[derive(Serialize, Deserialize, Debug)]
 pub struct JsonArray<T> {
-    flattened_vec: Vec<T>,
+    items: Vec<T>,
 }
 
 impl<T: de::DeserializeOwned> JsonArray<T> {
     pub fn from_response_str(str_to_parse: &str) -> Result<Self, ParseError> {
-        let mut parse_response = serde_json::from_str::<Value>(str_to_parse)?;
-        let parsed_json_array = parse_response
+        let items = try_parse_array_into_item_from_response_str(str_to_parse)?;
+        Ok(Self { items })
+    }
+
+    pub fn into_vec(self) -> Vec<T> {
+        self.items
+    }
+
+    pub fn from_json_array(json_value: &mut Value) -> Result<Self, ParseError> {
+        let array = json_value
             .as_array_mut()
             .ok_or_else(|| generate_none_error("cannot parse initial json string as array"))?;
 
-        let mut flattened_vec = Vec::new();
-        for nested_entry in parsed_json_array {
-            let parsed_item_value = match nested_entry.is_array() {
-                true => unwrap_item_in_nested_json_array(nested_entry.take())?,
-                false => nested_entry.take(),
-            };
-            let converted_item: T = serde_json::from_value(parsed_item_value)?;
-            flattened_vec.push(converted_item);
+        let mut items = Vec::new();
+        for item in array {
+            let converted_item = serde_json::from_value(item.take())?;
+            items.push(converted_item);
         }
+        Ok(Self { items })
+    }
 
-        Ok(Self { flattened_vec })
+    pub fn from_json_object(json_value: &mut Value) -> Result<Self, ParseError> {
+        let converted_object = serde_json::from_value(json_value.take())?;
+        let items = vec![converted_object];
+        Ok(Self { items })
     }
 
-    pub fn into_vec(self) -> Vec<T> {
-        self.flattened_vec
+    pub fn len(&self) -> usize {
+        self.items.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.items.is_empty()
```

Comment:
free utility operators that expose the inner `Vec` without having to actually move it